java/languages.antlr/src/org/netbeans/modules/languages/antlr/AntlrIndexer.java

```diff
@@ -0,0 +1,238 @@
+                    .forEach(s -> {
+                        id.addPair(FIELD_IMPORT, s, true, true);
+                    });
+        }
+
+        support.addDocument(id);
+    }
+
+    public static final class Factory extends EmbeddingIndexerFactory {
+
+        public static final String NAME = "antlr"; // NOI18N
+        public static final int VERSION = 4;
+        private static final int PRIORITY = 100;
+        private final String mimeType;
+
+        public Factory(String mimeType) {
+            this.mimeType = mimeType;
+        }
+
+        @Override
+        public EmbeddingIndexer createIndexer(final Indexable indexable, final Snapshot snapshot) {
+            if (isIndexable(indexable, snapshot)) {
+                return new AntlrIndexer();
+            } else {
+                return null;
+            }
+        }
+
+        @Override
+        public String getIndexerName() {
+            return NAME;
+        }
+
+        @Override
+        public int getIndexVersion() {
+            return VERSION;
+        }
+
+        private boolean isIndexable(Indexable indexable, Snapshot snapshot) {
+            return mimeType.equals(snapshot.getMimeType());
+        }
+
+        @Override
+        public void filesDeleted(Iterable<? extends Indexable> deleted, Context context) {
+            try {
+                IndexingSupport is = IndexingSupport.getInstance(context);
+                for (Indexable i : deleted) {
+                    is.removeDocuments(i);
+                }
+            } catch (IOException ioe) {
+                LOG.log(Level.WARNING, null, ioe);
+            }
+        }
+
+        @Override
+        public void rootsRemoved(final Iterable<? extends URL> removedRoots) {
+        }
+
+        @Override
+        public void filesDirty(Iterable<? extends Indexable> dirty, Context context) {
+            try {
+                IndexingSupport is = IndexingSupport.getInstance(context);
+                for (Indexable i : dirty) {
+                    is.markDirtyDocuments(i);
+                }
+            } catch (IOException ioe) {
+                LOG.log(Level.WARNING, null, ioe);
+            }
+        }
+
+        @Override
+        public int getPriority() {
+            return PRIORITY;
+        }
+    }
+
+
+    /**
+     * Find all imports of the specified file (including transitive ones).
+     *
+     * @param qs
+     * @param sourceFile
+     * @return
+     */
+    public static List<FileObject> transitiveImports(QuerySupport qs, FileObject sourceFile) {
+        List<FileObject> result = new ArrayList<>();
+        LinkedList<FileObject> toScan = new LinkedList<>();
+        Set<String> seen = new HashSet<>();
+        result.add(sourceFile);
+        toScan.add(sourceFile);
+        seen.add(sourceFile.getName());
+        while(! toScan.isEmpty()) {
+            FileObject target = toScan.poll();
+            try {
+                qs.getQueryFactory()
+                        .file(target)
```

Comment:
IllegalArgumentException can be thrown if the sourceFile does not belong to a project.